src/sdk/pynni/nni/bohb_advisor/bohb_advisor.py

```diff
@@ -0,0 +1,587 @@
+# Copyright (c) Microsoft Corporation
+# All rights reserved.
+#
+# MIT License
+#
+# Permission is hereby granted, free of charge,
+# to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction,
+# including without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and
+# to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+# BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+'''
+bohb_advisor.py
+'''
+
+from enum import Enum, unique
+import math
+import logging
+import json_tricks
+
+import ConfigSpace as CS
+import ConfigSpace.hyperparameters as CSH
+
+from nni.protocol import CommandType, send
+from nni.msg_dispatcher_base import MsgDispatcherBase
+
+from .config_generator import CG_BOHB
+
+logger = logging.getLogger('BOHB_Advisor')
+
+_next_parameter_id = 0
+_KEY = 'budget'
+_epsilon = 1e-6
+
+@unique
+class OptimizeMode(Enum):
+    """Optimize Mode class"""
+    Minimize = 'minimize'
+    Maximize = 'maximize'
+
+
+def create_parameter_id():
+    """Create an id
+
+    Returns
+    -------
+    int
+        parameter id
+    """
+    global _next_parameter_id  # pylint: disable=global-statement
+    _next_parameter_id += 1
+    return _next_parameter_id - 1
+
+
+def create_bracket_parameter_id(brackets_id, brackets_curr_decay, increased_id=-1):
+    """Create a full id for a specific bracket's hyperparameter configuration
+
+    Parameters
+    ----------
+    brackets_id: int
+        brackets id
+    brackets_curr_decay:
```

Comment:
brackets_curr_decay: float ?

Reply:
I lose this part... add already. It's int.